device/device.go

```diff
@@ -392,10 +407,8 @@ func (device *Device) SendKeepalivesToPeersWithCurrentKeypair() {
 
 	device.peers.RLock()
 	for _, peer := range device.peers.keyMap {
-		peer.keypairs.RLock()
-		sendKeepalive := peer.keypairs.current != nil && !peer.keypairs.current.created.Add(RejectAfterTime).Before(time.Now())
-		peer.keypairs.RUnlock()
-		if sendKeepalive {
+		current := peer.keypairs.Current()
+		if current.created.Add(RejectAfterTime).Before(time.Now()) {
```

Comment:
The nil check is now gone. Is that okay?

Also, wondering why this change was made.

Also, is this safe and race-free? It looks to me like created is now accessed without holding a lock? Maybe that's okay for other reasons?

Reply:
Good questions, yeah, I tried to look at this carefully, though it's been several months now.

The device carries 3 Keypairs pointers, current, previous and next. It used to use a mix of atomics and locks for accesses to these, but it was doing more work than strictly necessary in the critical path.

There is one field in Keypairs that is updated, which is sendNonce. That field is atomic. The other fields are initialized at creation time and otherwise remain unchanged - this means that the lock bound that we need is a lock bound around the adjustment of the pointers current, previous and next, but not around reading the contents of the destinations of those pointers (a win for GC).

Regarding the nil check, good catch. The reason this was missed is that there are no callers of this code in this repository, including tests, and Tailscale also doesn't call it. I see that the iOS bridge and client both call this. I'm adding a case to TestConcurrencySafety after bindupdate which is close to the iOS use case, it didn't trigger the NPE, but should nonetheless protect against bigger mistakes here.

Reply:
I followed up on this in #58. I'll merge this part through that PR, so you can refactor it out of here.